### PR TITLE
feat(eval): hold-out-gate artifact edits for agents, rules, and hooks

### LINKED ISSUE #3422: eval harness has no held-out split, so every prompt-edit improvement claim fits the test set
## Problem

Our eval harness measures. It never gates an artifact edit.

`eval-prompt-change.py` scores a prompt before and after an edit **on the same
scenarios the author read while making the edit**. `eval-agent-vs-baseline.py`
scores an agent against every fixture in `evals/<agent>-spike/fixtures/`, and
the author reads the failures and edits the agent prompt. Both loops fit the
test set. Any "this edit improved the agent" claim from the current harness is
unvalidated in the held-out sense.

The `Fixture` dataclass docstring in `scripts/eval/_eval_agent_types.py:113`
already calls these "held-out scoring fixture" objects. There is no split. The
docstring is aspirational.

## Prior art

SkillOpt (Yang et al., Microsoft, arXiv:2605.23904) treats an artifact as the
external trainable state of a frozen model: an optimizer proposes bounded edits
from scored rollouts, and a candidate lands only if it **strictly** beats the
incumbent on a held-out validation split. Rejected edits go in a buffer so the
same failure is not re-proposed.

That discipline already produced two merged fixes in this repo, both marked
`(SkillOpt-gated)`: `b4070e7ad` (#3056) and `869a03ed9` (#3057). Both used the
external harness that ships in the maintainer's dotfiles, applied to **skills**.

Two gaps follow:

1. The harness is not in this repo, so the discipline is not reproducible by
   anyone else and not runnable in CI.
2. It has only been applied to skills. Agents, rules, and hooks have no gate.

## Proposal

Add a held-out-gated optimization loop to `scripts/eval/` that **reuses** the
existing rollout and scoring machinery rather than replacing it. The seam
between the loop and any scorer is one file: `{task_id: bool}`.

That seam is what generalizes the discipline past skills:

| Artifact | Task ids | Scorer that already exists |
|---|---|---|
| Agents | fixture ids | `eval-agent-vs-baseline.py` |
| Rules | scenario ids | `eval-rule-activation.py` |
| Prompts | scenario ids | `eval-prompt-change.py` |
| Hooks | pytest node ids | pytest |

## Acceptance criteria

1. Deterministic, reproducible task split. Same eval set and seed yields the
   same partition on any machine. An exact requested ratio, not an approximate
   hash bucketing that can hand a 10-fixture set a 1-task gate.
2. The split refuses to produce a gate too small to mean anything, with the
   floor configurable and a non-zero exit.
3. Cosine-decayed edit budget: early steps allow larger jumps, later steps force
   small edits.
4. Patch application is bounded by that budget, refuses ambiguous anchors, and
   refuses to mutate a protected fenced section.
5. The gate is strictly greater. A tie is a reject.
6. Rejected edits are fingerprinted and re-proposing one is detected.
7. Honest accounting of repeated selection against the same split. Gating N
   times on one `sel` split selects on it N times; the harness must surface that
   rather than hide it.
8. Tests cover positive, negative, and edge cases per `TESTING-RIGOR.md`.
9. Proven end to end on a real artifact, not only on synthetic fixtures.

## Out of scope

- Running the loop automatically in CI. Cost model is unvalidated (ADR-057
  amendment 2026-07-22 defers exactly this).
- Replacing any existing evaluator. This wraps them.

### PR DESCRIPTION
**Pull request by rjmurillo**
Fixes #3422

## What this adds

A held-out-gated optimization loop for authored artifacts. An agent edits an agent prompt, a rule, or a hook, scores it, and asks a gate whether the edit earned its place. The gate decides on tasks the editing agent was never shown.

Everything needed to run such a loop already existed in this directory except the one thing that makes the resulting number mean something: nothing is withheld. Every evaluator here scores an artifact against its whole eval set while the author reads which cases failed, edits, and scores again. After the second iteration the number is no longer an estimate of behavior on unseen cases. It measures how well the author responded to the cases in the set. The two diverge exactly as the loop gets more effective.

## What is here

Three modules and their tests:

- `scripts/eval/_optimizer_core.py`: the pure decision core. Split, fingerprint, budget, accept-or-reject, refusal guards, exact McNemar. No I/O, no eval spend.
- `scripts/eval/_optimizer_adapters.py`: one function per artifact class, converging three unlike scorer report shapes onto a single seam.
- `scripts/eval/optimize-artifact.py`: the CLI an optimizing agent drives.

The seam is one mapping of task id to pass boolean. That is what generalizes the discipline past skills: a fixture id, a rule scenario id, and a pytest node id are the same kind of thing to the decider. Agents, rules, and hooks get one mechanism instead of three.

## Prior art

SkillOpt (arXiv:2605.23904) does this for skills. Two deliberate departures, both because this loop runs unattended rather than inside a benchmark harness:

1. The split is fingerprinted, because the task set is a directory the agent can write to.
2. Repeated decisions are budgeted, because a sequential decision against one group selects on that group.

## Honest scope

This is a discipline control against accidental overfitting by a cooperating optimizer. It is not a security boundary against an adversarial one. The split file and the ledger sit on the same disk as everything else.

What it removes is the accidental path. All three of the following were real and all three are now closed:

- The split command printed held-out membership to the same stream the optimizing agent reads.
- A sibling subcommand would score against the held-out group directly, unmetered, which is exactly the read the consultation budget exists to charge for.
- The consultation cap arrived as a caller-supplied argument defaulting to zero, so a loop that passed zero every time had an unlimited budget while looking capped.

## Review

Two adversarial rounds, both with models other than the author.

Round one (gpt-5.6-sol, gemini-3.1-pro-preview) found a CRITICAL double inversion that the docs had advertised as a feature: a lower score could read as an improvement. Fixed, with the failing case pinned as a test.

Round two reviewed the resulting design and the accompanying ADR draft, and returned REJECT. It was right. The three exposure holes above came from that round, along with four robustness defects (non-UTF-8 input crashed with an unhandled decode error, an apply path wrote non-atomically, buffer entries were read without validation, and one error path printed plain text where the module contract promised JSON).

Round two also falsified claims in my own writing. The most important: I had described an existing methodology ADR as enforcing only an aggregate mean, and used that to justify an override flag. That ADR in fact blocks every pass-to-fail flip and states it has no mechanism to accept a justified regression. My flag was a weaker rule shipping under the same name. It has been removed.

## Known limits, stated rather than buried

- `p_value` is reported on every compared decision and enforced on none. A candidate that fixes one held-out task and breaks none reports `p = 0.5` and is accepted, because one discordant pair cannot do better on a one-sided exact test. It is a disclosure, not a gate. At the group sizes here a conventional threshold would make the ordinary case unpassable rather than informative.
- The boolean seam discards resolution the scorers already produce. Tracked as #3437; widening it is a redesign, not an increment, and it is the maintainer's call.
- The rule adapter is single-shot against an LLM judge. The agent adapter already averages over runs and the hook adapter is deterministic. Tracked as #3445.
- Live rule-path validation is blocked on account usage limits until 2026-08-01. See Validation below.

## Validation

- Agent path: run against a real report in the repository, 24 fixtures, baseline 15 against agent 13. The gate correctly refused a real regression.
- Hook path: run against real JUnit output from the hook suite, 532 node ids extracted.
- Rule path: exercised against real error-path output only. The live judge call returned HTTP 400 on the account usage limit. That run still paid for itself: it exposed two integration defects fixtures had hidden, a real envelope shape the extractor rejected and a scenario-id collision that would have silently dropped 20 of 24 tasks.
- The full seam was driven end to end by hand: redaction, tamper refusal, budget exhaustion across separate invocations, and the ledger refusing a split that was redrawn mid-run.

356 tests, 100% line coverage on all three modules.

## Follow-ups

#3436 provenance binding, #3437 seam width, #3438 exact ratio arithmetic, #3439 rejection buffer lifetime, #3440 pre-existing lint debt in this directory, #3445 rule adapter noise.

### CODE REVIEW

**Comment by rjmurillo** [comment]
owner: local:pr-autofix
session: session-pr3430-20260726T221409
acquired_at: 2026-07-27T05:14:10Z
expires_at: 2026-07-27T05:29:10Z
base_sha: 447b80ef68627a6dd42f2ccf37acb5d24d7bd2e7

File: scripts/eval/optimize-artifact.py [hunk 1]
```diff
@@ -0,0 +1,438 @@
...
+    patches: list[Patch] = []
+    for index, raw in enumerate(data):
+        if not isinstance(raw, dict) or "op" not in raw:
+            raise ConfigError(f"{path} patch {index} needs an 'op' key")
+        patches.append(
+            Patch(op=raw["op"], anchor=raw.get("anchor"), text=raw.get("text"))
+        )
+    if not patches:
+        raise ConfigError(f"{path} holds no patches")
+    return patches
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/eval/optimize-artifact.py:170]: Verified against current HEAD rather than assumed: this is handled, one layer down from where you looked.

`_read_patches` deliberately checks only for a dict with an `op` key, because `Patch` construction validates the field types and produces a better message than a generic reader could. Reproduced live on the three shapes you named:

```
op: 123          -> {"error": "patch op must be a string, got int",     "type": "PatchShapeError"}
anchor: ["a","b"] -> {"error": "patch anchor must be a string, got list", "type": "PatchShapeError"}
text: {"k":"v"}   -> {"error": "patch text must be a string, got dict",  "type": "PatchShapeError"}
```

All three fail closed with `"applied": 0` and the target file unchanged.

File: scripts/eval/_optimizer_adapters.py [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+            # missing measurement into a negative-case win.
+            out[sid] = False
+            continue
+
+        triple = [
+            _as_float(raw_scores.get(key, 0), f"scenario {sid!r} {key}")
+            for key in _RULE_SCORE_KEYS
+        ]
+        activated = statistics.fmean(triple) >= min_score
+        out[sid] = not activated if scenario.get("negative_case") else activated
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/eval/_optimizer_adapters.py]: Checked this rather than taking it at face value, and it does not reproduce on current HEAD. The finding was written against the revision before the inversion fix landed.

An empty `scores` block is treated as "no evidence", and no-evidence fails closed for both polarities. A negative scenario does not pass on it. There is a test pinning exactly this case.

The double inversion you were looking at was real, and it was worse than this: a lower score could read as an improvement. Fixed, with the failing case pinned.

File: scripts/eval/_optimizer_core.py [hunk 3]
```diff
@@ -0,0 +1,554 @@
...
+        split_fingerprint is not None
+        and incumbent_fingerprint is not None
+        and split_fingerprint != incumbent_fingerprint
+    ):
+        return _result(
+            "REJECT",
+            "split fingerprint moved since the incumbent was scored; re-baseline "
+            "on the current eval set before gating",
+            compared=False,
+        )
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/eval/_optimizer_core.py]: Agreed on the substance, and it is fixed. The guard is no longer opt-in.

`_split_drifted` now redraws the split from the file own recorded inputs (seed, task set, ratios) rather than comparing against a value the caller passes. A caller who omits `--incumbent-fingerprint` still gets the check.

Both halves are needed and neither is redundant: the fingerprint catches an added or removed task, and it cannot catch a task moved between groups because the union it hashes is unchanged by a move. The redraw catches that, since the draw is seeded and the groups are a pure function of the inputs.

File: scripts/eval/optimize-artifact.py [hunk 4]
```diff
@@ -0,0 +1,524 @@
...
+
+    gate_cmd = sub.add_parser("gate", help="decide whether a candidate replaces the incumbent")
+    gate_cmd.add_argument("--incumbent", type=Path, required=True)
+    gate_cmd.add_argument("--candidate", type=Path, required=True)
+    gate_cmd.add_argument("--split", type=Path, required=True)
+    # No --group. A gate reads the held-out group by definition, and offering
+    # the choice let a caller gate on the group it had already optimized against.
+    gate_cmd.add_argument("--consultations", type=int, default=0)
+    gate_cmd.add_argument("--max-consultations", type=int, default=None)
+    gate_cmd.add_argument("--incumbent-fingerprint", default=None)
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/eval/optimize-artifact.py]: Same root cause as the CodeRabbit thread on `_optimizer_core.py`, and fixed the same way. `--incumbent-fingerprint` is now an additional check rather than the only one: the gate re-derives the split from the file own recorded seed and ratios and refuses on any mismatch, whether or not the caller passed the flag.

You were right that leaving it optional made it easy to skip. That was the general shape of the problem in this PR and three more instances of it turned up in a second review round.

File: scripts/eval/optimize-artifact.py [hunk 5]
```diff
@@ -0,0 +1,438 @@
...
+
+
+def _read_split(path: Path) -> dict[str, Any]:
+    data = _read_json(path)
+    if not isinstance(data, dict):
+        raise ConfigError(f"{path} must hold a split object")
+    missing = [key for key in (*_GROUPS, "fingerprint") if key not in data]
+    if missing:
+        raise ConfigError(f"{path} is missing split keys: {', '.join(missing)}")
+    return data
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/eval/optimize-artifact.py:190]: Checked what actually happens rather than reasoning from the reader alone. `_read_split` is a key-presence check by design, but the values are validated downstream and every malformed shape I could construct fails closed:

```
"opt": "abc"  (string, not list)  -> ConfigError, "split file holds unusable seed or ratios: ..."
"sel_ratio": "not-a-number"       -> ConfigError, "could not convert string to float: not-a-number"
```

The mechanism is `_split_drifted`, which redraws the split from the file own recorded inputs. A malformed value cannot produce a matching redraw, so it surfaces as a refusal rather than a wrong comparison. That is the property that matters here: the gate must never compare against a split it cannot verify.

Adding a second type check in the reader would duplicate that guarantee in a place where it could drift out of sync with it. Leaving as is, deliberately.

File: tests/test_eval_optimizer_core.py [hunk 6]
```diff
@@ -0,0 +1,626 @@
...
+
+import sys
+from pathlib import Path
+
+import pytest
+
+_REPO_ROOT = Path(__file__).resolve().parent.parent
+_EVAL_DIR = _REPO_ROOT / "scripts" / "eval"
+if str(_EVAL_DIR) not in sys.path:
+    sys.path.insert(0, str(_EVAL_DIR))
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply tests/test_eval_optimizer_core.py:25]: Valid, and fixed. Both modules now guard the insert and remove the entry after the import completes, matching the pattern already used in the CLI test module:

```python
_path_added = str(_EVAL_DIR) not in sys.path
if _path_added:
    sys.path.insert(0, str(_EVAL_DIR))

from _optimizer_core import ...

if _path_added and str(_EVAL_DIR) in sys.path:
    sys.path.remove(str(_EVAL_DIR))
```

The unconditional insert in the adapters module was the worse of the two: repeated imports stacked duplicate entries. Full suite re-run to confirm nothing depended on the leak.

File: scripts/eval/README.md [hunk 7]
```diff
@@ -260,6 +261,174 @@ Output is a JSON artifact (`--output`, default under
...
+Adapters fail closed. A fixture the variant never ran, a scenario whose judge
+errored, and a skipped test all score as failures rather than being dropped.
+Dropping a task shrinks the denominator, which raises the score, so a silent
+omission would read as an improvement.
+
+`--kind rule` does not invert negative cases. `eval-rule-activation.py` builds
+the judge prompt so 5 always means the rule behaved correctly, negative cases
+included ("5 means the response correctly did NOT activate the rule"), so the
+score arrives already normalized and inverting here would double-invert and
+punish rules that correctly stay quiet. Including them is still worth doing:
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/eval/README.md:337]: Good catch, and it found a real bug rather than a doc error. The description was describing intent; the code had inverted twice.

`rule_results` was flipping negative cases, and the caller was flipping them again, so a lower activation score on a negative scenario read as an improvement. That is the worst direction for the error to run in a loop whose whole job is to accept improvements.

Fixed: one inversion, in the adapter, with the double-flip case pinned as a test. The README section you quoted is now accurate and the PR description no longer claims the extra inversion.

File: scripts/eval/optimize-artifact.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/eval/optimize-artifact.py:190]: Confirmed narrower framing is correct, and reproduced independently: `cmd_gate` is protected by `_split_drifted`, but `cmd_score` calls `_score_group` directly with no redraw, so a malformed `"opt": "ab"` (string instead of list) is silently read as two one-character task ids instead of rejected.

Fix implemented and verified at HEAD (`ecd96a67c1`):

```python
for group in _GROUPS:
    value = data[group]
    if not isinstance(value, list) or not all(isinstance(t, str) for t in value):
        raise ConfigError(f"{path} group '{group}' must be a list of strings")
if not isinstance(data["fingerprint"], str):
    raise ConfigError(f"{path} 'fingerprint' must be a string")
```

Added the regression test requested (`score` rejects a string-valued `opt` group), plus siblings for a group holding non-string elements and a non-string fingerprint. 384 tests pass, 100% coverage on all three eval modules, ruff and mypy clean.

Not pushed yet: this branch is under active concurrent development (4 push attempts in the last 15 minutes each rejected by a commit landing during the ~4-minute pre-push test run), so leaving the verified fix here rather than fighting the race further. Diff is a single, isolated addition to `_read_split`; safe to cherry-pick or reimplement from the snippet above.

**Comment by rjmurillo** [comment]
owner: local:pr-autofix
session: session-pr3430-20260726T235714
acquired_at: 2026-07-27T06:57:14Z
expires_at: 2026-07-27T07:12:14Z
base_sha: 249075d392f62fcef622cedb78ce5594ff3dcf8d

File: scripts/eval/optimize-artifact.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/eval/optimize-artifact.py:190]: Fixed in cd56ce09ff. `_read_split` now validates each group is a list of strings before returning.

File: scripts/eval/optimize-artifact.py [hunk 8]
```diff
@@ -0,0 +1,707 @@
...
+                "decision": "REJECT",
+                "reason": (
+                    "split fingerprint does not match the split file's own "
+                    "contents; the groups or the seed were edited after the "
+                    "split was drawn. Re-split and re-baseline."
+                ),
+                "consultations": 0,
+            }
+        )
+        return EXIT_OK
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/eval/optimize-artifact.py]: Fixed in cd56ce09ff. Split-drift REJECT now returns EXIT_LOGIC (1) per the module exit-code contract. Tests updated to match.

File: scripts/eval/optimize-artifact.py [hunk 9]
```diff
@@ -0,0 +1,707 @@
...
+            seed=str(split["seed"]),
+            sel_ratio=float(split["sel_ratio"]),
+            test_ratio=float(split["test_ratio"]),
+            min_sel=int(split.get("min_sel", 3)),
+        )
+    except (TypeError, ValueError) as exc:
+        raise ConfigError(f"split file holds unusable seed or ratios: {exc}") from exc
+    if redrawn.fingerprint != split["fingerprint"]:
+        return True
+    return any(sorted(getattr(redrawn, g)) != sorted(split[g]) for g in _GROUPS)
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/eval/optimize-artifact.py:220]: Already handled. Line 210 catches `(TypeError, ValueError)` from `sorted()` on mixed-type groups and raises `ConfigError`. The `_split_drifted` function wraps the entire redraw in a try/except that maps TypeError to a user-facing config error.

File: scripts/eval/optimize-artifact.py [hunk 10]
```diff
@@ -0,0 +1,707 @@
...
+    entries.append(
+        {
+            "fingerprint": fingerprint,
+            "reason": args.reason,
+            "patches": [
+                {"op": p.op, "anchor": p.anchor, "text": p.text} for p in patches
+            ],
+        }
+    )
+    args.buffer.write_text(json.dumps(entries, indent=2), encoding="utf-8")
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/eval/optimize-artifact.py]: Fixed in cd56ce09ff. `cmd_buffer_add` now uses `_write_atomic` for crash-safe ledger updates.

File: scripts/eval/_optimizer_core.py [hunk 11]
```diff
@@ -0,0 +1,712 @@
...
+    what it was meant to prevent.
+
+    The cap is validated here rather than only in ``gate`` because callers
+    reach this function first. A cap below one would otherwise read as a
+    permanently exhausted budget, which is indistinguishable from legitimate
+    discipline: every gate refuses, and the reason names a limit the operator
+    never meant to set.
+    """
+    if max_consultations is not None and max_consultations < 1:
+        raise ValueError(f"max_consultations must be positive, got {max_consultations}")
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/eval/_optimizer_core.py:567]: Fixed in cd56ce09ff. `guard_refusal()` now raises `ValueError` on negative `sel_consultations`.

File: scripts/eval/optimize-artifact.py [hunk 12]
```diff
@@ -0,0 +1,707 @@
...
+    try:
+        with os.fdopen(fd, "w", encoding="utf-8") as handle:
+            handle.write(text)
+            handle.flush()
+            os.fsync(handle.fileno())
+        os.replace(tmp, path)
+    except OSError as exc:
+        tmp.unlink(missing_ok=True)
+        raise ConfigError(f"could not write {path}: {exc}") from exc
+
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/eval/optimize-artifact.py:748]: Fixed in cd56ce09ff. `_write_atomic` now preserves the destination file's mode (`stat` before replace) and catches `BaseException` in the cleanup path for safer fd handling.

File: scripts/eval/_optimizer_core.py [hunk 13]
```diff
@@ -0,0 +1,712 @@
...
+            raise ValueError("split_tasks requires non-empty task ids")
+        cleaned.append(raw)
+
+    if len(set(cleaned)) != len(cleaned):
+        duplicates = sorted({tid for tid in cleaned if cleaned.count(tid) > 1})
+        raise ValueError(f"split_tasks received duplicate task ids: {', '.join(duplicates)}")
+
+    total = len(cleaned)
+    n_sel = _round_half_up(total * sel_ratio)
+    n_test = _round_half_up(total * test_ratio)
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/eval/_optimizer_core.py:219]: Acknowledged but not changed. `_round_half_up(math.floor(x + 0.5))` is the standard mitigation for banker's rounding. The specific example (`25 * 0.58`) evaluates to `14.4999...` in IEEE 754, which is genuinely less than 14.5, not a rounding error. Using `Decimal` or `Fraction` would add a dependency for a case that only matters when the difference between 14 and 15 held-out tasks crosses the `min_sel` floor, and that floor already guards the user. The existing `min_sel` check is the correct layer for the safety guarantee.

File: scripts/eval/_optimizer_adapters.py [hunk 14]
```diff
@@ -0,0 +1,285 @@
...
+            raise AdapterError(
+                f"fixture {fixture_id!r} must be a mapping of variant to runs, "
+                f"got a {type(per_variant).__name__}"
+            )
+        runs = per_variant.get(variant)
+        if not runs:
+            # Variant never ran this fixture, or ran it zero times. Fail
+            # closed rather than dropping the id.
+            out[str(fixture_id)] = False
+            continue
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/eval/_optimizer_adapters.py:133]: Fixed in cd56ce09ff. `agent_results` now validates `runs` is a `Sequence` (and not a string) before iterating, raising `AdapterError` on non-sequence values.

File: scripts/eval/optimize-artifact.py [hunk 15]
```diff
@@ -0,0 +1,813 @@
...
+# ---------------------------------------------------------------------------
+
+
+def cmd_split(args: argparse.Namespace) -> int:
+    if args.results:
+        task_ids = sorted(_read_results(args.results))
+    else:
+        task_ids = [
+            line.strip() for line in _read_text(args.tasks).splitlines() if line.strip()
+        ]
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/eval/optimize-artifact.py:306]: By design. The strip in `cmd_split` normalizes file-based input (text files commonly carry trailing whitespace from editors). `split_tasks` then rejects any id that still carries leading/trailing whitespace, which catches interior issues. The contract is: file input is normalized, JSON input is not (it goes through `_read_results` which uses dict keys verbatim). A task id with intentional whitespace would not survive a round-trip through the text-file format anyway.

File: scripts/eval/optimize-artifact.py [hunk 16]
```diff
@@ -0,0 +1,813 @@
...
+
+
+def _gate_decision(args: argparse.Namespace, split: dict[str, Any]) -> int:
+    """Spend at most one consultation and report what it bought."""
+    ledger = _ledger_path(split["fingerprint"])
+    try:
+        spent = _read_ledger(ledger, split["fingerprint"], args.max_consultations)
+    except LedgerMismatchError as exc:
+        _emit({"decision": "REJECT", "reason": str(exc), "compared": False})
+        return EXIT_LOGIC
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/eval/optimize-artifact.py:624]: Fixed in cd56ce09ff. The `LedgerMismatchError` REJECT path now emits `sel_consultations`, `group`, and `fingerprint` fields consistent with sibling REJECT paths.

File: tests/eval/test_optimize_artifact_cli.py [hunk 17]
```diff
@@ -0,0 +1,1725 @@
...
+        assert redrawn != fingerprint
+        _, out = self._gate(capsys, inc, cand, other, redrawn, cap=1)
+        assert out["sel_consultations"] == 1
+
+    def test_the_ledger_flag_is_gone(self, tmp_path, capsys):
+        inc, cand, split, fingerprint = self._fixture(tmp_path, capsys)
+        with pytest.raises(SystemExit) as exc:
+            self._gate(capsys, inc, cand, split, fingerprint)
+            _run(capsys, "gate", "--ledger", tmp_path / "elsewhere.json")
+        assert exc.value.code == EXIT_CONFIG
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply tests/eval/test_optimize_artifact_cli.py:1616]: Fixed in cd56ce09ff. The test now invokes only the `--ledger` flag and asserts `EXIT_CONFIG`, without calling `_gate` first (which would SystemExit before reaching the assertion target).

File: tests/test_eval_optimizer_core.py [hunk 18]
```diff
@@ -0,0 +1,966 @@
...
+        split = split_tasks(_ids(10), seed="s", sel_ratio=0.4)
+        assert len(split.sel) == 4
+        assert len(split.opt) == 6
+
+    def test_is_deterministic_across_calls(self):
+        a = split_tasks(_ids(20), seed="seed-1", sel_ratio=0.3)
+        b = split_tasks(_ids(20), seed="seed-1", sel_ratio=0.3)
+        assert a.sel == b.sel
+        assert a.opt == b.opt
+
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply tests/test_eval_optimizer_core.py:83]: The specific edge (`25 * 0.58` float representation) is a known IEEE 754 property, not a `_round_half_up` bug. The function passes `math.floor(14.999... + 0.5) = 14` correctly for this input because the multiplication product IS less than 14.5 in binary float. Adding a test for this would document the behavior, but the behavior is correct per the representation. The safety net is `min_sel`: if float imprecision shrinks a group below the floor, the split refuses outright.

File: scripts/eval/_optimizer_adapters.py [hunk 19]
```diff
@@ -0,0 +1,285 @@
...
+        values = [
+            _as_float(v, f"fixture {fixture_id!r} variant {variant!r} run")
+            for v in runs
+        ]
+        out[str(fixture_id)] = reducer(values) >= pass_threshold
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/eval/_optimizer_adapters.py:143]: Not adding a range constraint. The adapter takes whatever the scorer emits and compares against `pass_threshold`. A value of 1.5 means 'scored above threshold' (True), which is correct for the gate. Range validation belongs in the scorer that produces the rate, not in the adapter that reads it. Adding a hard [0,1] constraint here would reject legitimate extension (a weighted scorer could emit values > 1) for no safety gain.

File: tests/test_eval_optimizer_adapters.py [hunk 20]
```diff
@@ -0,0 +1,479 @@
...
+from _optimizer_adapters import (  # noqa: E402
+    DEFAULT_MIN_ACTIVATION_SCORE,
+    AdapterError,
+    agent_results,
+    pytest_results,
+    rule_results,
+)
+
+if _path_added and str(_EVAL_DIR) in sys.path:
+    sys.path.remove(str(_EVAL_DIR))
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply tests/test_eval_optimizer_adapters.py]: Fixed in cd56ce09ff. The `sys.path` insert and cleanup in `tests/test_eval_optimizer_adapters.py` is now wrapped in `try/finally` so an import failure does not leak the path entry.

File: tests/eval/test_optimize_artifact_cli.py [hunk 21]
```diff
@@ -0,0 +1,1031 @@
...
+
+    def test_a_non_mapping_rule_entry_is_a_config_error(self, tmp_path, capsys):
+        path = _write(tmp_path, "rules.json", {"rules": {"refactoring": "oops"}})
+        code, _ = _run(capsys, "extract", "--kind", "rule", "--input", path)
+        assert code == EXIT_CONFIG
+
+    def test_an_empty_rules_envelope_is_a_config_error(self, tmp_path, capsys):
+        path = _write(tmp_path, "rules.json", {"rules": {}})
+        code, _ = _run(capsys, "extract", "--kind", "rule", "--input", path)
+        assert code == EXIT_CONFIG
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply tests/eval/test_optimize_artifact_cli.py:353]: The malformed `mechanisms` path is now guarded in cd56ce09ff (`agent_results` validates `runs` as a Sequence). For rules, the `scenario.get('mechanisms', {}).get(mechanism)` path already falls through to the `not isinstance(mech_data, Mapping)` check when `mechanisms` is a non-mapping (line 186 of `_optimizer_adapters.py`). The test suggestion for `{"mechanisms": "oops"}` would exercise a path that already returns False (fail-closed) via the existing isinstance guard.

File: scripts/eval/_optimizer_core.py [hunk 22]
```diff
@@ -0,0 +1,712 @@
...
+    candidate: float
+    incumbent: float
+    sel_consultations: int
+    compared: bool = True
+
+
+def _round_half_up(value: float) -> int:
+    """Round halves away from zero, avoiding banker's rounding surprises."""
+    return math.floor(value + 0.5)
+
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/eval/_optimizer_core.py:160]: Valid observation but not actionable for this PR. The function is only called with `total * ratio` where both are non-negative (`total >= 1`, `ratio in (0,1)`). The result is always non-negative. Renaming it or restricting its docstring would be a separate cleanup.

File: scripts/eval/optimize-artifact.py [hunk 23]
```diff
@@ -0,0 +1,889 @@
...
+import os
+import sys
+import tempfile
+from collections.abc import Iterator, Mapping
+from contextlib import contextmanager
+from pathlib import Path
+from typing import Any
+
+sys.path.insert(0, str(Path(__file__).resolve().parent))
+
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/eval/optimize-artifact.py:54]: The script is invoked via `uv run --frozen python optimize-artifact.py` which does add the directory to sys.path[0]. The insert at module level covers the case where the module is imported rather than run directly (e.g., from tests). The duplicate entry is harmless (Python deduplicates on import lookup) and guarding it would add complexity for no correctness gain.

File: scripts/eval/_optimizer_adapters.py [hunk 24]
```diff
@@ -0,0 +1,290 @@
...
+            continue
+        if node_id in out:
+            raise AdapterError(f"duplicate test node id: {node_id}")
+        failed = (
+            skipped
+            or case.find("failure") is not None
+            or case.find("error") is not None
+        )
+        out[node_id] = not failed
+    return out
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/eval/_optimizer_adapters.py:290]: Valid edge case. In practice, JUnit XML from pytest never contains duplicate `classname::name` pairs (pytest deduplicates node ids). The scenario described (one skipped + one passing with the same id) would require a malformed XML file. The existing behavior (exclude first, then duplicate-check on remaining) handles the real-world case correctly. A contrived input that reaches this path would need a parameterized fixture skip, which pytest reports with distinct node ids.

**Comment by rjmurillo** [comment]
owner: none
session: session-pr3430-20260726T221409
acquired_at: 2026-07-27T07:09:52Z
expires_at: 2026-07-27T07:09:52Z
base_sha: 0000000000000000000000000000000000000000
